README.md
<p align="center">
    <picture >
      <source media="(prefers-color-scheme: light)" srcset="https://brand.papirfly.com/fr/gallery/34433/images/lowres/24ced107-4a60-45ef-8be5-c14cfe133908.png">
      <source media="(prefers-color-scheme: dark)" srcset="https://brand.papirfly.com/fr/gallery/34433/images/lowres/00a0679f-ece6-41a0-8039-be5ec0fe6700.png">
      <img alt="Shows the papirfly logo" src="https://brand.papirfly.com/fr/gallery/34433/images/lowres/24ced107-4a60-45ef-8be5-c14cfe133908.png" height=100>
    </picture>
</p>
<h2 align="center">Papirfly authenticator</h2>
<p align="center">
<strong>Authorize your secure web app with a Papirfly brand portal.</strong>
<br><br>

[![Maintenance Status][maintenance-image]](#maintenance-status)

### Why you may want to use this library

The Papirfly authenticator is a OAuth client implementation, that allows you to quickly integrate towards a 
papirfly brand portal, either through using the [Authorization code flow](https://www.rfc-editor.org/rfc/rfc6749#section-1.3.1) or the [Client credentials flow](https://www.rfc-editor.org/rfc/rfc6749#section-1.3.4).

The Papirfly authenticator also supports the [PKCE](https://tools.ietf.org/html/rfc7636) ("Pixy") extension to OAuth.

Note: While this client will work towards most OAuth2 Servers, it officially only supports Papirfly Brand Portal.

## Supported methods

### `authorizeWithAuthorizationCode`

This is the main and preferred function to use for authentication. Invoking this function will do the whole login
flow, opening a popup and returns the access token, refresh token and access token expiry date when successful, or it
throws an error when not successful.

```ts
import { authorizeWithAuthorizationCode } from "@papirfly/papirfly-authenticator";

const config = {
    grantType: "authorization_code",
    clientId: "<YOUR_CLIENT_ID>",
    scopes: ["<YOUR_SCOPES_ARRAY>"],
    serviceConfiguration: {
        authorizationEndpoint: "<FULL_URL_TO_AUTHORIZATION_ENDPOINT>",
        tokenEndpoint: "<FULL_URL_TO_TOKEN_ENDPOINT>",
    },
    redirectConfiguration: {
        url: "<YOUR_REDIRECT_URL>",
        authorizedMessage: "<NAME_OF_SUCCSESSFUL_AUTHORIZATION_MESSAGE>",
        rejectedMessage: "<NAME_OF_REJECTED_AUTHORIZATION_MESSAGE>",
    },
};

const result = await authorizeWithAuthorizationCode(config);
```
### `authorizeWithClientCredentials`

This method allows you to authenticate without using a popup. Invoking this function will do the whole login
flow and returns the access token, refresh token and access token expiry date when successful, or it
throws an error when not successful.

```ts
import { authorizeWithClientCredentials } from "@papirfly/papirfly-authenticator";

const config = {
    grantType: "client_credentials",
    clientId: "<YOUR_CLIENT_ID>",
    clientSecret: "<YOUR_CLIENT_SECRET>",
    scopes: ["<YOUR_SCOPES_ARRAY>"],
    serviceConfiguration: {
        authorizationEndpoint: "<FULL_URL_TO_AUTHORIZATION_ENDPOINT>",
        tokenEndpoint: "<FULL_URL_TO_TOKEN_ENDPOINT>",
    },
};

const result = await authorizeWithClientCredentials(config);
```


### `refresh`

This method will refresh the accessToken using the refreshToken.

```ts
import { refresh } from "@papirfly/papirfly-authenticator";

const config = { "<CONFIG_USED_DURING_INITIAL_AUTHORIZATION>"};

const result = await refresh(config, {
    refreshToken: `<REFRESH_TOKEN>`,
});
```

## Getting started

```sh
npm install @papirfly/papirfly-authenticator --save
```

### Error messages

You can catch errors by adding a `throwError` function in your configurations. This will return a Error object and the configuration used to cause the error.

#### Maintenance Status

**Active:** Papirfly is actively working on this project, and we expect to continue working on it for the foreseeable future.
Bug reports, feature requests and pull requests are welcome!

[maintenance-image]: https://img.shields.io/badge/maintenance-active-green.svg


CHANGELOG.md

<a name="0.0.2"></a>
## 0.0.2 (2022-11-14)
### Chores
- add changelog

<a name="0.0.1"></a>
## 0.0.1 (2022-11-14)
### Features
- First implementation


__tests__/OAuth2Utils.test.ts
import * as sinon from "sinon";
import { PapirflyAuthenticator, PapirflyAuthenticatorInterfaces } from "../src";
describe("PapirflyAuthenticator", () => {
    const emptyAuthFlowConfig: PapirflyAuthenticatorInterfaces.IAuthConfiguration = {
        grantType: "authorization_code",
        clientId: "",
        clientSecret: "",

        scopes: [],
        serviceConfiguration: {
            authorizationEndpoint: "",
            tokenEndpoint: "",
        },
        redirectConfiguration: {
            url: "someurl",
            authorizedMessage: "",
            rejectedMessage: "",
        },
    };
    afterEach(() => {
        sinon.restore();
    });

    it("should throw error if popup fails to open", async () => {
        const error = sinon.stub();
        //given
        sinon.stub(window, "open");
        //when
        await PapirflyAuthenticator.authorizeWithAuthorizationCode({ ...emptyAuthFlowConfig, throwError: error });
        //then
        expect(error.callCount).toBe(1);
    });

    it("should throw error if popup unexpectedly closes", async () => {
        const error = sinon.stub();
        //given
        sinon
            .stub(window, "open")
            .returns({ location: { replace: sinon.stub() }, closed: true, close: sinon.stub() } as any);
        //when
        await PapirflyAuthenticator.authorizeWithAuthorizationCode({ ...emptyAuthFlowConfig, throwError: error });
        //then
        expect(error.callCount).toBe(1);
    });

    it("should try to authorize when refresh fails", async () => {
        //given
        sinon.stub(PapirflyAuthenticator, "fetchToken").throws();
        const authorize = sinon.stub(PapirflyAuthenticator, "authorizeWithAuthorizationCode");
        //when
        await PapirflyAuthenticator.refresh(emptyAuthFlowConfig, { refreshToken: "", grantType: "refresh_token" });
        //then
        expect(authorize.callCount).toBe(1);
    });

    it("should try to authorize with credentials when refresh fails", async () => {
        //given
        sinon.stub(PapirflyAuthenticator, "fetchToken").throws();
        const authorize = sinon.stub(PapirflyAuthenticator, "authorizeWithClientCredentials");
        //when
        await PapirflyAuthenticator.refresh(
            {
                grantType: "client_credentials",
                clientId: "",
                clientSecret: "",
                scopes: [],
                serviceConfiguration: {
                    authorizationEndpoint: "",
                    tokenEndpoint: "",
                },
            },
            { refreshToken: "", grantType: "refresh_token" }
        );
        //then
        expect(authorize.callCount).toBe(1);
    });

    it("should create token expiration date", () => {
        //given
        const expiresIn1Hour = 3600;
        //when
        const expirationDate = PapirflyAuthenticator.getAccessTokenExpirationDate(expiresIn1Hour);
        //then
        expect(expirationDate).toBeGreaterThan(Date.now());
    });

    it("should create token expiration date when no expiry date is given", () => {
        //given
        const noExpiry: number | undefined = undefined;
        //when
        const expirationDate = PapirflyAuthenticator.getAccessTokenExpirationDate(noExpiry);
        //then
        expect(expirationDate).toBeGreaterThan(Date.now());
    });

    it("should create random code challenges", async () => {
        //when
        const challenge1 = await PapirflyAuthenticator.createChallenge();
        const challenge2 = await PapirflyAuthenticator.createChallenge();

        expect(challenge1).not.toEqual(challenge2);
    });
});


src/OAuth2UtilsTypes.ts
export namespace OAuth2UtilsTypes {
    export interface IServiceConfiguration {
        /**
         * Fully formed url to the OAuth authorization endpoint.
         *
         * Usually ends with `/oauth/authorize`
         */
        authorizationEndpoint: string;
        /**
         * Fully formed url to the OAuth token exchange endpoint.
         *
         * Usually ends with `/oauth/token`
         */
        tokenEndpoint: string;
    }

    export interface IBaseConfiguration {
        /**
         * Configuration for token exchange endpoints.
         */
        serviceConfiguration: IServiceConfiguration;
        /**
         * Catches any potential errors
         */
        throwError?: ((error: unknown, config: IBaseAuthConfiguration) => Promise<void>) | undefined;
    }

    interface IBaseAuthGrant {
        grantType: OAuthGrantType;
    }

    interface IRedirectConfiguration {
        /**
         *  URL the authorization server will redirect the user back to the application a user successfully authorizes an application.
         */
        url: string;
        /**
         * Name of message sent from redirect window on succsessful authorization.
         */
        authorizedMessage: string;
        /**
         * Name of message sent from redirect window on rejected authorization.
         */
        rejectedMessage: string;
    }

    export interface IBaseAuthConfiguration extends IBaseConfiguration, IBaseAuthGrant {
        /**
         * Public identifier for app.
         */
        clientId: string;
    }

    export interface IAuthConfiguration extends IBaseAuthConfiguration {
        grantType: "authorization_code";
        /**
         * Secret known only to the application and the authorization server. Should never be commited.
         *
         * Optional for authorization code flow.
         */
        clientSecret?: string;
        /**
         * Scope to control access and help the user identify the permissions they are granting to the application.
         */
        scopes: string[];

        /**
         * Configuration for redirect.
         * Redirect should send an Event of type "message", where the authorization code is found under code
         */
        redirectConfiguration: IRedirectConfiguration;
        /**
         * Enable Proof Key for Code Exchange (PKCE, pronounced Pixy).
         *
         * Default `true`.
         */
        usePKCE?: boolean;
    }

    export interface IAuthWithCredentialsConfiguration extends IBaseAuthConfiguration {
        grantType: "client_credentials";
        /**
         * Secret known only to the application and the authorization server.
         *
         * Mandatory for client credentials flow
         */
        clientSecret: string;
        /**
         * Scope to control access and help the user identify the permissions they are granting to the application.
         */
        scopes?: string[];
        redirectUrl?: never;
    }

    interface IBaseResult {
        /**
         * Token used for API requests.
         */
        accessToken: string;
        /**
         * Expiration date of token in milliseconds.
         */
        accessTokenExpirationDate: number;
        /**
         * Token used to refresh accesstoken when it has expired.
         */
        refreshToken?: string;
        /**
         *  Representation of how an access_token will be generated and presented for resource access calls.
         *
         *  Usually `Bearer`.
         */
        tokenType: string;
    }

    export interface IAuthorizeResult extends IBaseResult {
        /**
         * Scope of accessToken.
         */
        scopes: string[];
        authorizationCode?: string;
        codeVerifier?: string;
    }

    export interface IRefreshResult extends IBaseResult {}

    export interface IRefreshConfiguration extends IBaseAuthGrant {
        grantType: "refresh_token";
        /**
         * Token received from authorization/previous refresh.
         */
        refreshToken: string;
    }

    export interface ICodeChallenge {
        /**
         * Cryptographically random string used to verify challenge on the OAuth endpoint.
         */
        verifyer: string;
        /**
         * Cryptographically random string derived from verifyer used to verify requests on the OAuth endpoint.
         */
        challenge: string;
    }

    export interface IAuthError {
        error: string;
        error_description: string;
    }

    export type OAuthGrantType = "client_credentials" | "refresh_token" | "password" | "authorization_code";

    // https://tools.ietf.org/html/rfc6749#section-4.1.2.1
    type OAuthAuthorizationErrorCode =
        | "unauthorized_client"
        | "access_denied"
        | "unsupported_response_type"
        | "invalid_scope"
        | "server_error"
        | "temporarily_unavailable";
    // https://tools.ietf.org/html/rfc6749#section-5.2
    type OAuthTokenErrorCode =
        | "invalid_request"
        | "invalid_client"
        | "invalid_grant"
        | "unauthorized_client"
        | "unsupported_grant_type"
        | "invalid_scope";
    type AppAuthErrorCode = "popup_closed_unexpectedly" | "popup_not_opened";

    export type ErrorCode = OAuthAuthorizationErrorCode | OAuthTokenErrorCode | AppAuthErrorCode;
}


src/PapirflyAuthenticator.ts
import { PapirflyAuthenticatorInterfaces as oauth } from "./PapirflyAuthenticatorInterfaces";

/**
 * Class containing utilities for authorizing a application towards 3rd party integrations with OAuth2
 */
export class PapirflyAuthenticator {
    private static challengeMethod = { algorithm: "SHA-256", key: "S256" };

    /**
     * Authorize client with Authorization Code flow
     */
    static async authorizeWithAuthorizationCode(
        config: oauth.IAuthConfiguration
    ): Promise<oauth.IAuthorizeResult | undefined> {
        const urlParams = [
            ["response_type", "code"],
            ["client_id", config.clientId],
            ["scope", config.scopes.join("+")],
            ["redirect_uri", config.redirectConfiguration.url],
        ];

        let codeChallenge: oauth.ICodeChallenge | undefined;
        if (config.usePKCE !== false) {
            codeChallenge = await this.createChallenge();
            urlParams.push(
                ["code_challenge_method", this.challengeMethod.key],
                ["code_challenge", codeChallenge.challenge]
            );
        }
        const encodedUrlParams = urlParams
            .map(function (p) {
                return p[0] + "=" + p[1];
            })
            .join("&");
        const serviceConfiguration = config.serviceConfiguration;
        const url = `${serviceConfiguration.authorizationEndpoint}?${encodedUrlParams}`;
        const popup = open(url, "Authorize", "top=100,left=100");
        if (!popup) {
            config.throwError?.(new Error("popup_not_opened" as oauth.ErrorCode), config);
            return;
        }
        const listenForAuthenticationCode = new Promise<string>((resolve, reject) => {
            let popupHasBeenClosed = false;
            const listener = (event: MessageEvent<any>) => {
                const message = JSON.parse(event.data);
                if (message.name == config.redirectConfiguration.authorizedMessage) {
                    resolve(message.code);
                    closePopup();
                } else if (message.name == config.redirectConfiguration.rejectedMessage) {
                    reject({ error: message.error, error_description: message.error_description });
                    closePopup();
                }
            };
            addEventListener("message", listener);

            const closePopup = () => {
                if (popup) {
                    popup.close();
                }
                if (listener) {
                    removeEventListener("message", listener);
                }
                if (interval) {
                    clearInterval(interval);
                }
                popupHasBeenClosed = true;
            };
            const interval = setInterval(() => {
                if (popup && popup.closed && !popupHasBeenClosed) {
                    popupHasBeenClosed = true;
                    closePopup();
                    reject({ error: "popup_closed_unexpectedly", error_description: "popup_closed_unexpectedly" });
                }
            }, 1 * 500);
        });

        try {
            const authorizationCode = await listenForAuthenticationCode;
            const token = await this.getToken(authorizationCode, "authorization_code", config, codeChallenge?.verifyer);
            return {
                accessToken: token.access_token,
                refreshToken: token.refresh_token,
                scopes: config.scopes,
                accessTokenExpirationDate: this.getAccessTokenExpirationDate(token.expires_in),
                authorizationCode: authorizationCode,
                tokenType: token.token_type,
            };
        } catch (error) {
            config.throwError?.(error, config);
            return;
        }
    }

    /**
     * Authorize client with Client Credentials flow
     */
    static async authorizeWithClientCredentials(
        config: oauth.IAuthWithCredentialsConfiguration
    ): Promise<oauth.IAuthorizeResult | undefined> {
        try {
            const token = await this.getToken(config.clientSecret, "client_credentials", config);
            return {
                accessToken: token.access_token,
                refreshToken: token.refresh_token,
                scopes: config.scopes || [],
                accessTokenExpirationDate: this.getAccessTokenExpirationDate(token.expires_in),
                tokenType: token.token_type,
            };
        } catch (error) {
            config.throwError?.(error, config);
            return;
        }
    }

    /**
     * Refresh OAuth session
     */
    static refresh: (
        config: oauth.IAuthConfiguration | oauth.IAuthWithCredentialsConfiguration,
        refreshConfig: oauth.IRefreshConfiguration
    ) => Promise<oauth.IRefreshResult | undefined> = async (config, refreshConfig) => {
        const body = new FormData();

        body.append("grant_type", refreshConfig.grantType);
        body.append("refresh_token", refreshConfig.refreshToken);
        body.append("client_id", config.clientId);
        if (config.clientSecret) body.append("client_secret", config.clientSecret);

        try {
            const serviceConfiguration = config.serviceConfiguration;
            const token = await this.fetchToken(serviceConfiguration.tokenEndpoint, body);
            return {
                accessToken: token.access_token,
                refreshToken: token.refresh_token,
                tokenType: token.token_type,
                accessTokenExpirationDate: this.getAccessTokenExpirationDate(token.expires_in),
            };
        } catch (error) {
            //If refresh fails, try to reauthorize
            if (config.grantType === "authorization_code") {
                return this.authorizeWithAuthorizationCode(config);
            } else {
                return this.authorizeWithClientCredentials(config);
            }
        }
    };

    static getAccessTokenExpirationDate(expires_in: number | undefined) {
        return Math.min(
            // Add a bit safety threshold to avoid using an expired access token
            Date.now() + (expires_in || 3600) * 0.9 * 1000,
            // Refresh token at minimum once per day
            Date.now() + 24 * 3600 * 1000
        );
    }

    private static getToken(
        authorizationCode: string,
        grantType: oauth.OAuthGrantType,
        config: oauth.IAuthConfiguration | oauth.IAuthWithCredentialsConfiguration,
        verifyer?: string
    ) {
        const serviceConfiguration = config.serviceConfiguration;
        const body = new FormData();

        body.append("grant_type", grantType);
        body.append("code", authorizationCode);
        body.append("client_id", config.clientId);
        if (config.grantType === "authorization_code") body.append("redirect_uri", config.redirectConfiguration.url);
        if (config.scopes) body.append("scope", config.scopes.join("+"));
        if (config.clientSecret) body.append("client_secret", config.clientSecret);
        if (verifyer) body.append("code_verifier", verifyer);

        return this.fetchToken(serviceConfiguration.tokenEndpoint, body);
    }

    static async fetchToken(tokenEndpoint: string, body: FormData) {
        const response = await fetch(tokenEndpoint, {
            method: "POST",
            body: body,
        });
        if (response.status !== 200) {
            const errorText = await response.text();
            const errorMessage = `${response.status}: ${errorText})}`;
            throw new Error(errorMessage);
        }

        const json = await response.json();
        return json as {
            access_token: string;
            token_type: string;
            expires_in: number;
            refresh_token: string;
        };
    }

    /**
     * Create code challenge for PKCE (Proof Key for Code Exchange).
     * Challenge is sent in initial authorization such that subsequent requests that include the verifyer are validated.
     * Used to prevent malicious attacks by replacing the static secret as it is exposed in the browser.
     */
    static async createChallenge(): Promise<oauth.ICodeChallenge> {
        const allChars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
        const randomInts = new Uint8Array(50);
        crypto.getRandomValues(randomInts);
        const chars = randomInts.map(function (x) {
            return allChars.charCodeAt(x % allChars.length);
        });
        const verifyer = String.fromCharCode.apply(null, Array.from(chars));

        const digest = await crypto.subtle.digest(this.challengeMethod.algorithm, chars);
        const hash = String.fromCharCode.apply(null, Array.from(new Uint8Array(digest)));
        return { challenge: btoa(hash).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, ""), verifyer: verifyer };
    }
}

export const authorizeWithAuthorizationCode = (config: oauth.IAuthConfiguration) =>
    PapirflyAuthenticator.authorizeWithAuthorizationCode(config);
export const authorizeWithClientCredentials = (config: oauth.IAuthWithCredentialsConfiguration) =>
    PapirflyAuthenticator.authorizeWithClientCredentials(config);
export const refresh = (
    config: oauth.IAuthConfiguration | oauth.IAuthWithCredentialsConfiguration,
    refreshConfig: oauth.IRefreshConfiguration
) => PapirflyAuthenticator.refresh(config, refreshConfig);


src/OAuth2Utils.ts
import { OAuth2UtilsTypes as oauth } from "./OAuth2UtilsTypes";

/**
 * Class containing utilities for authorizing a application towards 3rd party integrations with OAuth2.
 * * Use `authorizeWithAuthorizationCode` if you need to access user specific files. Will open a pop-up where the user will login.
 * * Use `authorizeWithClientCredentials` if you do not need to access user specific files.
 * * After authorizing you can refresh the access tokens with `refresh`.
 * Authorization code flow is preferred.
 * @static throwError - Function for any potential errors.
 */
export class OAuth2Utils {
    private static challengeMethod = { algorithm: "SHA-256", key: "S256" };

    /**
     * Authorize client with Authorization Code flow
     */
    static async authorizeWithAuthorizationCode(
        config: oauth.IAuthConfiguration
    ): Promise<oauth.IAuthorizeResult | undefined> {
        const urlParams = [
            ["response_type", "code"],
            ["client_id", config.clientId],
            ["scope", config.scopes.join("+")],
            ["redirect_uri", config.redirectConfiguration.url],
        ];

        let codeChallenge: oauth.ICodeChallenge | undefined;
        if (config.usePKCE !== false) {
            codeChallenge = await this.createChallenge();
            urlParams.push(
                ["code_challenge_method", this.challengeMethod.key],
                ["code_challenge", codeChallenge.challenge]
            );
        }
        const encodedUrlParams = urlParams
            .map(function (p) {
                return p[0] + "=" + p[1];
            })
            .join("&");
        const serviceConfiguration = config.serviceConfiguration;
        const url = `${serviceConfiguration.authorizationEndpoint}?${encodedUrlParams}`;
        const popup = open(url, "Authorize", "top=100,left=100");
        if (!popup) {
            config.throwError?.(new Error("popup_not_opened" as oauth.ErrorCode), config);
            return;
        }
        const listenForAuthenticationCode = new Promise<string>((resolve, reject) => {
            let popupHasBeenClosed = false;
            const listener = (event: MessageEvent<any>) => {
                const message = JSON.parse(event.data);
                if (message.name == config.redirectConfiguration.authorizedMessage) {
                    resolve(message.code);
                    closePopup();
                } else if (message.name == config.redirectConfiguration.rejectedMessage) {
                    reject({ error: message.error, error_description: message.error_description });
                    closePopup();
                }
            };
            addEventListener("message", listener);

            const closePopup = () => {
                if (popup) {
                    popup.close();
                }
                if (listener) {
                    removeEventListener("message", listener);
                }
                if (interval) {
                    clearInterval(interval);
                }
                popupHasBeenClosed = true;
            };
            const interval = setInterval(() => {
                if (popup && popup.closed && !popupHasBeenClosed) {
                    popupHasBeenClosed = true;
                    closePopup();
                    reject({ error: "popup_closed_unexpectedly", error_description: "popup_closed_unexpectedly" });
                }
            }, 1 * 500);
        });

        try {
            const authorizationCode = await listenForAuthenticationCode;
            const token = await this.getToken(authorizationCode, "authorization_code", config, codeChallenge?.verifyer);
            return {
                accessToken: token.access_token,
                refreshToken: token.refresh_token,
                scopes: config.scopes,
                accessTokenExpirationDate: this.getAccessTokenExpirationDate(token.expires_in),
                authorizationCode: authorizationCode,
                tokenType: token.token_type,
            };
        } catch (error) {
            config.throwError?.(error, config);
            return;
        }
    }

    /**
     * Authorize client with Client Credentials flow
     */
    static async authorizeWithClientCredentials(
        config: oauth.IAuthWithCredentialsConfiguration
    ): Promise<oauth.IAuthorizeResult | undefined> {
        try {
            const token = await this.getToken(config.clientSecret, "client_credentials", config);
            return {
                accessToken: token.access_token,
                refreshToken: token.refresh_token,
                scopes: config.scopes || [],
                accessTokenExpirationDate: this.getAccessTokenExpirationDate(token.expires_in),
                tokenType: token.token_type,
            };
        } catch (error) {
            config.throwError?.(error, config);
            return;
        }
    }

    /**
     * Refresh OAuth session
     */
    static refresh: (
        config: oauth.IAuthConfiguration | oauth.IAuthWithCredentialsConfiguration,
        refreshConfig: oauth.IRefreshConfiguration
    ) => Promise<oauth.IRefreshResult | undefined> = async (config, refreshConfig) => {
        const body = new FormData();

        body.append("grant_type", refreshConfig.grantType);
        body.append("refresh_token", refreshConfig.refreshToken);
        body.append("client_id", config.clientId);
        if (config.clientSecret) body.append("client_secret", config.clientSecret);

        try {
            const serviceConfiguration = config.serviceConfiguration;
            const token = await this.fetchToken(serviceConfiguration.tokenEndpoint, body);
            return {
                accessToken: token.access_token,
                refreshToken: token.refresh_token,
                tokenType: token.token_type,
                accessTokenExpirationDate: this.getAccessTokenExpirationDate(token.expires_in),
            };
        } catch (error) {
            //If refresh fails, try to reauthorize
            if (config.grantType === "authorization_code") {
                return this.authorizeWithAuthorizationCode(config);
            } else {
                return this.authorizeWithClientCredentials(config);
            }
        }
    };

    static getAccessTokenExpirationDate(expires_in: number | undefined) {
        return Math.min(
            // Add a bit safety threshold to avoid using an expired access token
            Date.now() + (expires_in || 3600) * 0.9 * 1000,
            // Refresh token at minimum once per day
            Date.now() + 24 * 3600 * 1000
        );
    }

    private static getToken(
        authorizationCode: string,
        grantType: oauth.OAuthGrantType,
        config: oauth.IAuthConfiguration | oauth.IAuthWithCredentialsConfiguration,
        verifyer?: string
    ) {
        const serviceConfiguration = config.serviceConfiguration;
        const body = new FormData();

        body.append("grant_type", grantType);
        body.append("code", authorizationCode);
        body.append("client_id", config.clientId);
        if (config.grantType === "authorization_code") body.append("redirect_uri", config.redirectConfiguration.url);
        if (config.scopes) body.append("scope", config.scopes.join("+"));
        if (config.clientSecret) body.append("client_secret", config.clientSecret);
        if (verifyer) body.append("code_verifier", verifyer);

        return this.fetchToken(serviceConfiguration.tokenEndpoint, body);
    }

    static async fetchToken(tokenEndpoint: string, body: FormData) {
        const response = await fetch(tokenEndpoint, {
            method: "POST",
            body: body,
        });
        if (response.status !== 200) {
            const errorText = await response.text();
            const errorMessage = `${response.status}: ${errorText})}`;
            throw new Error(errorMessage);
        }

        const json = await response.json();
        return json as {
            access_token: string;
            token_type: string;
            expires_in: number;
            refresh_token: string;
        };
    }

    /**
     * Create code challenge for PKCE (Proof Key for Code Exchange).
     * Challenge is sent in initial authorization such that subsequent requests that include the verifyer are validated.
     * Used to prevent malicious attacks by replacing the static secret as it is exposed in the browser.
     */
    static async createChallenge(): Promise<oauth.ICodeChallenge> {
        const allChars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
        const randomInts = new Uint8Array(50);
        crypto.getRandomValues(randomInts);
        const chars = randomInts.map(function (x) {
            return allChars.charCodeAt(x % allChars.length);
        });
        const verifyer = String.fromCharCode.apply(null, Array.from(chars));

        const digest = await crypto.subtle.digest(this.challengeMethod.algorithm, chars);
        const hash = String.fromCharCode.apply(null, Array.from(new Uint8Array(digest)));
        return { challenge: btoa(hash).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, ""), verifyer: verifyer };
    }
}

export const authorizeWithAuthorizationCode = (config: oauth.IAuthConfiguration) =>
    OAuth2Utils.authorizeWithAuthorizationCode(config);
export const authorizeWithClientCredentials = (config: oauth.IAuthWithCredentialsConfiguration) =>
    OAuth2Utils.authorizeWithClientCredentials(config);
export const refresh = (
    config: oauth.IAuthConfiguration | oauth.IAuthWithCredentialsConfiguration,
    refreshConfig: oauth.IRefreshConfiguration
) => OAuth2Utils.refresh(config, refreshConfig);
